Mathlib/Lean/Meta/RefinedDiscrTree/Basic.lean

```diff
@@ -1,178 +1,266 @@
-
-instance : LT Key := ⟨fun a b => Key.lt a b⟩
-instance (a b : Key) : Decidable (a < b) := inferInstanceAs (Decidable (Key.lt a b))
-
 private def Key.format : Key → Format
-  | .star i                 => "*" ++ Std.format i
+  | .star id                => f!"*{id}"
   | .opaque                 => "◾"
-  | .const k a              => "⟨" ++ Std.format k ++ ", " ++ Std.format a ++ "⟩"
-  | .fvar k a               => "⟨" ++ Std.format k.name ++ ", " ++ Std.format a ++ "⟩"
-  | .lit (Literal.natVal v) => Std.format v
-  | .lit (Literal.strVal v) => repr v
-  | .sort                   => "sort"
-  | .bvar i a               => "⟨" ++ "#" ++ Std.format i ++ ", " ++ Std.format a ++ "⟩"
+  | .const name nargs       => f!"⟨{name}, {nargs}⟩"
+  | .fvar fvarId nargs      => f!"⟨{fvarId.name}, {nargs}⟩"
+  | .lit (Literal.natVal n) => f!"{n}"
+  | .lit (Literal.strVal s) => f!"{s.quote}"
+  | .sort                   => "Sort"
+  | .bvar i nargs           => f!"⟨#{i}, {nargs}⟩"
   | .lam                    => "λ"
   | .forall                 => "∀"
-  | .proj s i a             => "⟨" ++ Std.format s ++"."++ Std.format i ++", "++ Std.format a ++ "⟩"
+  | .proj name idx nargs    => f!"⟨{name}.{idx}, {nargs}⟩"
 
 instance : ToFormat Key := ⟨Key.format⟩
 
+/--
+Converts an entry (i.e., `List Key`) to the discrimination tree into
+`MessageData` that is more user-friendly.
+-/
+partial def keysAsPattern (keys : Array Key) : CoreM MessageData := do
+  let (msg, keys) ← go (paren := false) |>.run keys.toList
+  if !keys.isEmpty then
+    throwError "illegal discrimination tree entry: {keys.map Key.format}"
+  return msg
+where
+  /-- Get the next key. -/
+  next : StateRefT (List Key) CoreM Key := do
+    let key :: keys ← get | throwError "illegal discrimination tree entry: {keys.map Key.format}"
+    set keys
+    return key
+  /-- Format the application `f args`. -/
+  mkApp (f : MessageData) (nargs : Nat) (paren : Bool) : StateRefT (List Key) CoreM MessageData :=
+    if nargs == 0 then
+      return f
+    else do
+      let mut r := f
+      for _ in [:nargs] do
+        r := r ++ m!" {← go}"
+      return parenthesize m!"{r}" paren
+
+  /-- Format the next expression. -/
+  go (paren := true) : StateRefT (List Key) CoreM MessageData := do
+    let key ← next
+    match key with
+    | .const declName nargs =>
+      mkApp m!"{← mkConstWithLevelParams declName}" nargs paren
+    | .fvar fvarId nargs =>
+      mkApp m!"{mkFVar fvarId}" nargs paren
+    | .proj _ i nargs =>
+      mkApp m!"{← go}.{i+1}" nargs paren
+    | .bvar i nargs =>
+      mkApp m!"#{i}" nargs paren
+    | .lam =>
+      return parenthesize m!"λ, {← go (paren := false)}" paren
+    | .forall =>
+      return parenthesize m!"{← go} → {← go (paren := false)}" paren
+    | _ => return key.format
+  /-- Add parentheses if `paren == true`. -/
+  parenthesize (msg : MessageData) (paren : Bool) : MessageData :=
+    if paren then m! "({msg})" else msg
```

Comment:
```suggestion
    if paren then msg.paren else msg
```
which knows about line wrapping. You might want to use `.group` in the else branch for line-wrapping reasons too.

Reply:
I've upstreamed this to https://github.com/leanprover/lean4/pull/7200